archivist_samples/document/document.py

```diff
@@ -0,0 +1,176 @@
+            "arc_attribute_type": "arc_attachment",
+            "arc_blob_identity": attachment["identity"],
+            "arc_blob_hash_alg": attachment["hash"]["alg"],
+            "arc_blob_hash_value": attachment["hash"]["value"],
+            "arc_display_name": attachment_description.attribute_name,
+            "arc_file_name": attachment_description.filename,
+        }
+
+        return result
+
+
+document_creator = make_assets_create(
+    attachment_creator=attachment_create, confirm=True, public=True
+)
+
+
+class Document:
+    selector_key = "OnboardingSample"
+    selector_value = "DocumentLineage"
+
+    def __init__(
+        self,
+        arch: "type_helper.Archivist",
+        display_type: str,
+    ):
+        arch_ = copy(arch)
+        arch_.fixtures = {
+            "assets": {
+                "attributes": {
+                    "arc_display_type": display_type,
+                },
+            },
+        }
+
+        self._arch = arch_
+        self._asset = None
+        self._existed = False
+
+    @property
+    def arch(self):
+        return self._arch
+
+    @property
+    def asset(self):
+        return self._asset
+
+    @property
+    def existed(self):
+        return self._existed
+
+    def create(
+        self,
+        name: str,
+        description: str,
+        *,
+        attachments: Optional[list] = None,
+        custom_attrs: Optional[dict] = None,
+    ):
+        attrs = {
+            "arc_description": description,
+            "arc_profile": "Document",
+        }
+
+        if custom_attrs is not None:
+            attrs.update(custom_attrs)
+
+        self._asset, self._existed = document_creator(
+            self.arch,
+            name,
+            attrs,
+            attachments=attachments,
+            selector_key=self.selector_key,
+            selector_value=self.selector_value,
+        )
+
+        return self._asset
+
+    # Publish new version of the document
+    def publish(
+        self,
+        document: dict,
+        version: str,
+        doc_hash: str,
+        authors: List[Dict],
+        name: str,
+    ):
+        props = {
+            "operation": "Record",
+            "behaviour": "RecordEvidence",
+        }
+        attrs = {
+            "arc_display_type": "Publish",
+            "document_version_authors": authors,
+        }
+
+        asset_attrs = {
+            "arc_display_name": name,
+            "document_document": document,
+            "document_hash_value": doc_hash,
+            "document_hash_alg": "sha256",
+            "document_version": version,
+            "document_status": "Published",
+        }
+
+        return self.arch.events.create(
+            self.asset["identity"],
+            props=props,
+            attrs=attrs,
+            asset_attrs=asset_attrs,
+            confirm=True,
+        )
+
+    # Withdraw version of the document
+    def withdraw(self, document: dict, version: str, hash: str, name: str):
```

Comment:
hash is a builtin and this will fail checks

Reply:
renamed to doc_hash!